### PR TITLE
review pmd check rules

### PR DESCRIPTION
**Pull request by alievmirza**
Let's select rules that we want to add to our `check-rules/pmd-rules.xml`. 
I propose to review at least ruels with `<priority>1</priority>` and `<priority>2</priority>` in errorprone.xml and multithreading.xml 

Original set of rules: https://github.com/pmd/pmd/blob/master/pmd-java/src/main/resources/category/java/errorprone.xml

### CODE REVIEW

File: check-rules/errorprone.xml [hunk 1]
```diff
@@ -0,0 +1,3721 @@
...
+        <example>
+            <![CDATA[
+public class Foo {
+   protected void finalize() {}
+}
+]]>
+        </example>
+    </rule>
+
+    <rule name="EmptyFinallyBlock"
```
Comment (alievmirza) [inline check-rules/errorprone.xml:1620]: This rule is enabled in our repo

File: check-rules/errorprone.xml [hunk 2]
```diff
@@ -0,0 +1,3721 @@
...
+            <![CDATA[
+void foo() {
+    Bar b = new Bar();
+    b.finalize();
+}
+]]>
+        </example>
+    </rule>
+
+    <rule name="AvoidCatchingNPE"
```
Comment (alievmirza) [inline check-rules/errorprone.xml:213]: This rule is enabled in our repo

File: check-rules/errorprone.xml [hunk 3]
```diff
@@ -0,0 +1,3721 @@
...
+}
+
+private int getMoreFoo(){
+      return this.moreFoo;
+}
+]]>
+        </example>
+    </rule>
+
+    <rule name="BrokenNullCheck"
```
Comment (alievmirza) [inline check-rules/errorprone.xml:674]: This rule is enabled in our repo

File: check-rules/errorprone.xml [hunk 4]
```diff
@@ -0,0 +1,3721 @@
...
+      // this is probably not what you meant to do
+      ;
+      // the extra semicolon here this is not necessary
+      System.out.println("look at the extra semicolon");;
+}
+]]>
+        </example>
+    </rule>
+
+    <rule name="EmptySwitchStatements"
```
Comment (alievmirza) [inline check-rules/errorprone.xml:1793]: This rule is enabled in our repo

File: check-rules/errorprone.xml [hunk 5]
```diff
@@ -0,0 +1,3721 @@
...
+      { _bar = bar; } // Why not?
+      {} // But remove this.
+   }
+
+}
+]]>
+        </example>
+    </rule>
+
+    <rule name="EmptyStatementNotInLoop"
```
Comment (alievmirza) [inline check-rules/errorprone.xml:1750]: This rule is enabled in our repo

File: check-rules/errorprone.xml [hunk 6]
```diff
@@ -0,0 +1,3721 @@
...
+    switch (x) {
+        // once there was code here
+        // but it's been commented out or something
+    }
+}
+]]>
+        </example>
+    </rule>
+
+    <rule name="EmptySynchronizedBlock"
```
Comment (alievmirza) [inline check-rules/errorprone.xml:1822]: This rule is enabled in our repo

File: check-rules/errorprone.xml [hunk 7]
```diff
@@ -0,0 +1,3721 @@
...
+public boolean test(String s) {
+    if (s == "one") return true;        // unreliable
+    if ("two".equals(s)) return true;   // better
+    return false;
+}
+]]>
+        </example>
+    </rule>
+
+    <rule name="UselessOperationOnImmutable"
```
Comment (alievmirza) [inline check-rules/errorprone.xml:3602]: This rule is enabled in our repo

File: check-rules/bestpractices.xml [hunk 8]
```diff
@@ -0,0 +1,2130 @@
...
+        <example>
+            <![CDATA[
+import java.io.File;  // not referenced or required
+import java.util.*;   // not referenced or required
+public class Foo {}
+]]>
+        </example>
+    </rule>
+
+    <rule name="UnusedLocalVariable"
```
Comment (alievmirza) [inline check-rules/bestpractices.xml:1638]: This rule is enabled in our repo

File: check-rules/bestpractices.xml [hunk 9]
```diff
@@ -0,0 +1,2130 @@
...
+public class Foo {
+    public void doSomething() {
+        int i = 5; // Unused
+    }
+}
+]]>
+        </example>
+    </rule>
+
+    <rule name="UnusedPrivateField"
```
Comment (alievmirza) [inline check-rules/bestpractices.xml:1660]: This rule is enabled in our repo

File: check-rules/errorprone.xml [hunk 10]
```diff
@@ -0,0 +1,3721 @@
...
+    // this would trigger the rule (and throw a ClassCastException if executed)
+Integer[] a = (Integer [])c.toArray();
+
+   // this is fine and will not trigger the rule
+Integer[] b = (Integer [])c.toArray(new Integer[c.size()]);
+]]>
+        </example>
+    </rule>
+
+    <rule name="CloneMethodMustBePublic"
```
Comment (zstan) [inline check-rules/errorprone.xml:864]: I stopped here, seems all upper is appropriate and very helpful.

**Pull request closed without merging**